bin/combinators.ml

```diff
@@ -1,14 +1,59 @@
-Random.init 2025
+(* random int generators are from qcheck 
+  - allows test results to be consistent
+  - shortcut for the function signature 
+  I'll probably create my own version tho *)
 
-let rec int_list s = 
+let rec int_list_size_gen s st = 
   if s <= 0 then
     []
   else
-    Random.int 100000 :: int_list (s - 1)
+    QCheck.Gen.int st :: int_list_size_gen (s - 1) st
 
-let rec print_list l = match l with 
-| [] -> print_newline ()
-| h ::t -> print_int h; print_char ' '; print_list t
+(* int list sorted in ascending order *)
+let rec int_list_sorted_gen prev s st = 
+  if s <= 0 then
+    []
+  else
+    let n = QCheck.Gen.int st in
+    if n >= prev then
+      n :: int_list_sorted_gen n (s - 1)st
+    else 
+      int_list_sorted_gen prev (s - 1) st
+
+(* int list with each element identical *)
+let int_list_dup_gen s st = 
+  let n = QCheck.Gen.int st in
+  let rec aux s' st' =
+    if s' <= 0 then
+      []
+    else
+        n :: aux (s' - 1) st' in
+  aux s st
+
+(* int list with each element unique *)
+(* either dup or unique needs more work *)
+let int_list_unique_gen s st = 
+  let set = Hashtbl.create s in
+  let rec aux s' st' =
+    if s' <= 0 then
+      []
+    else
+      let n = QCheck.Gen.int st' in
+      (fun x -> Hashtbl.replace set x ()) n;
+      if s' <> Hashtbl.length set then 
+        n :: aux (s' - 1) st' else 
+        aux (s' - 1) st' in
```

Comment:
This seems a little strange to me. Maybe you can clarify. So the length of set is growing on each iteration, the size of s' is shrinking on each iteration, and the only time they are equal would be one time when the paths cross? I don't think you want to use s' here and instead need to compute the length of the list each time(Or just check for membership). 

What am I missing atm?